### DISCUSSION TITLE
slint::invoke_from_event_loop SegFault (C++)

### DISCUSSION THREAD

**Discussion by gammaEmitter** [slint-ui/slint#2365] [Q&A] [answered]
Hello!

I'm calling a function in a thread that gets the `slint::ComponentHandle<MainApp>` as an argument. The function itself works as intended. When the function returns, the program terminates with a SegFault (unfortunately the MacOS debugger doesn't show up anymore, I don't have further description and am not trained to use another debugger ^^).

I narrowed it down to the `slint::invoke_from_event_loop` causing the SegFault. It is no UAF nor any access to `StreamInformation` Members that is faulty. 

Is this because the  `std::shared_ptr` being held by `slint::ComponentHandle<MainApp>` is freed at the end of the function scope ? If so, how can I implement my function without that happening?

```
struct PlotMeterArguments
{
    slint::ComponentHandle<HelloWorld> * App;
    InputStreamValues * streaminformation;
};
```

`pthread_create(&Helper::meter_thread, NULL,PlotMeter,(void*)&arguments);`

```
void * PlotMeter (void * args) {
    pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS,NULL);

    PlotMeterArguments * arguments = (PlotMeterArguments *) args;

    slint::ComponentHandle<HelloWorld> * AppHandle = (slint::ComponentHandle<HelloWorld> *) arguments->App;
    InputStreamValues * StreamInformation = (InputStreamValues * ) arguments->streaminformation;

    Pa_Sleep(250); //wait until other thread opened the stream
    while (StreamInformation->err == Pa_IsStreamActive(StreamInformation->stream))
    {
        
        for (int i = 0; i < FRAMES_PER_BUFFER; i++)
        {
           
            slint::invoke_from_event_loop([&](){(*AppHandle)->fn_set_meter_to_value(0.5f);}); 
        }

    }

    return 0;
}
```

**Comment by ogoffart** [comment]
> Is this because the std::shared_ptr being held by slint::ComponentHandle<MainApp> is freed at the end of the function scope ? If so, how can I implement my function without that happening?

Probably.  
Instead of taking a pointer to `slint::ComponentHandle<HelloWorld> * `, you should  just take the ComponentHandle by value (``slint::ComponentHandle<HelloWorld>` without the `*`) to make sure the handle stays alive.

**Reply by gammaEmitter** [reply]
Well, passing `slint::ComponentHandle<HelloWorld>` by value I need to use `slint::ComponentHandle<HelloWorld> App = HelloWorld::create();` as the constructor, because there is no default constructor (compared to the pointer version. 
This leads me to the same point that the function is working but the App shuts down after the function returns.

A workaround with passing a `slint::ComponentWeakHandle<HelloWorld>` then calling `lock()` ends in a `type slint::ComponentWeakHandle<HelloWorld>` that does not have the member function anymore.

**Reply by ogoffart** [reply]
Calling `lock()`  should give back a `slint::ComponentHandle<HelloWorld>`

**Comment by gammaEmitter** [comment, marked as answer]
- Calling `lock()` returned a `std::optional<slint::ComponentHandle<HelloWorld>>` which my function could not been called from (Member does not exist)
- Calling `lock().__get()` returned the correct `slint::ComponentHandle<HelloWorld>` which I could call my function from, yet it terminated the app, wanting me to call the function from `slint::invoke_from_event_loop` (_dispite I did call by it_)

Wrong Lines:
```
slint::ComponentHandle<HelloWorld> * AppHandle = (slint::ComponentHandle<HelloWorld> *) arguments->App;
slint::invoke_from_event_loop([&](){(*AppHandle)->fn_set_meter_to_value(0.5f);}); 
```

Correct Solution: I was not passing `AppHandle` in the lambdas capture clause
```
slint::ComponentHandle<HelloWorld> AppHandle = (slint::ComponentHandle<HelloWorld>) arguments->App;
slint::invoke_from_event_loop([AppHandle](){AppHandle->fn_set_meter_to_value(0.5f);}); 
```

Thanks again for staying with me!
